### PR TITLE
Double Dispatch technique

### PR DESCRIPTION
**Pull request by Morheit**

### CODE REVIEW

File: test/ConnectionTest.cpp [hunk 1]
```diff
@@ -15,6 +15,16 @@ namespace {
         return connection.advancedQuery().serverVersion;
     }
 
+    int getDispatchedConnectionInfo(const IDBConnection2& dbConnection) {
+        QueryReceiver queryReceiver;
```
Comment (malirod) [inline test/ConnectionTest.cpp]: @Morheit, I suggest to follow Google coding style. In particular naming conventions: https://google.github.io/styleguide/cppguide.html#Naming

File: src/DBConnection.hpp [hunk 2]
```diff
@@ -23,4 +23,35 @@ class MySqlDBConnection : public IDBConnection {
...
+};
+
+class MySqlDBConnection2 : public IDBConnection2 {
+public:
+    explicit MySqlDBConnection2(std::string serverVersion, int protocolVersion);
+
+    int query() const;
+    Info advancedQuery() const;
+
+private:
```
[the remark this answers is not in the document]
Comment (Morheit) [reply src/DBConnection.hpp]: @malirod , purpose is to divide private members and private methods

File: src/DBConnection.hpp [hunk 3]
```diff
@@ -23,4 +23,35 @@ class MySqlDBConnection : public IDBConnection {
 
 private:
     Info info;
+};
+
+class QueryReceiver;
+
+class IDBConnection2 {
```
[the remark this answers is not in the document]
Comment (Morheit) [reply src/DBConnection.hpp]: Done

File: src/DBConnection.hpp [hunk 4]
```diff
@@ -23,4 +23,35 @@ class MySqlDBConnection : public IDBConnection {
...
+public:
+    virtual ~IDBConnection2() = default;
+
+    virtual int sendQuery(const QueryReceiver& connectionVisitor) const = 0;
+    virtual Info sendAdvancedQuery(const QueryReceiver& connectionVisitor) const = 0;
```
[the remark this answers is not in the document]
Comment (Morheit) [reply src/DBConnection.hpp]: @malirod , `sendAdvancedQuery` removed. Added new implementation of Double Dispatch. Base class doesn't knows about it's children anymore

File: src/DBConnection2.hpp [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    MySqlDBConnection2() = default;
+    explicit MySqlDBConnection2(std::string serverVersion, int protocolVersion);
+
+    void dispatch(const ConnectionDispatcher& connectionDispatcher) const override;
+    int query() const override;
+
+    Info advancedQuery() const;
+
+private:
+    Info info;
```
Comment (malirod) [inline src/DBConnection2.hpp]: @Morheit , usually private members has special prefixes or suffixes. I propose "m_info" or "info_"

File: src/ConnectionDispatcher.hpp [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+class ConnectionDispatcher {
+public:
+    ConnectionDispatcher() = default;
+    ~ConnectionDispatcher();
+
+    void dispatchConnection(const MySqlDBConnection2& connection) const;
+    const MySqlDBConnection2& getMySqlConnection() const;
+
+private:
+    mutable std::unique_ptr<const MySqlDBConnection2> m_connection;
```
Comment (malirod) [inline src/ConnectionDispatcher.hpp]: @Morheit , I see to much of "const" and mutable. Are you sure you need this?

File: src/DBConnection2.hpp (hunk 5 above)
Comment (Morheit) [reply src/DBConnection2.hpp]: @malirod ok. `info` replaced by `m_info`

File: src/ConnectionDispatcher.hpp (hunk 6 above)
Comment (Morheit) [reply src/ConnectionDispatcher.hpp]: @malirod I doing this way to clarificate that we don't need any changes on this object and we need only use of functions

File: test/ConnectionTest.cpp (hunk 1 above)
Comment (Morheit) [reply test/ConnectionTest.cpp]: @malirod Done

File: src/DBConnection2.cpp [hunk 7]
```diff
@@ -0,0 +1,20 @@
+#include <iostream>
```
Comment (malirod) [inline src/DBConnection2.cpp]: @Morheit , I see no iostream usage

File: src/ConnectionDispatcher.hpp [hunk 8]
```diff
@@ -0,0 +1,17 @@
+#pragma once
+
+#include <memory>
+
+#include "DBConnection2.hpp"
+
+class ConnectionDispatcher {
```
Comment (malirod) [inline src/ConnectionDispatcher.hpp]: @Morheit ,
1) I don't see the reason for storing ptr in smart ptr. you can store just ptr.
2) see http://qaru.site/questions/89999/what-is-single-and-double-dispatch. I don't see the point to store ptr at all. In the sample you can see just sequence of calls and that's it. There is just "magic" with parameters and thats it. Try to implement something like this. Consider passing lambdas to dispatcher to pass custom hander.

File: test/ConnectionTest2.cpp [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        dbConnection.dispatch(connectionDispatcher);
+        return connectionDispatcher.getMySqlConnection().advancedQuery().serverVersion;
+    }
+
+    TEST(Connection2InfoTest, GetInfo) {
+        MySqlDBConnection2 connection("4.2.2 MySQL Server", 10);
+        ASSERT_EQ(10, getConnection2Info(connection));
+    }
+
+    TEST(Connection2InfoTest, GetAdvancedInfo) {
```
Comment (malirod) [inline test/ConnectionTest2.cpp]: @Morheit , consider test case with some SQLiteDBConnection which has only query().

Comment (Morheit) [reply test/ConnectionTest2.cpp]: @malirod , test case with SQLiteDBConnection added

File: src/ConnectionDispatcher.hpp (hunk 8 above)
Comment (Morheit) [reply src/ConnectionDispatcher.hpp]: @malirod smart ptr replaced by ptr

File: src/DBConnection2.cpp (hunk 7 above)
Comment (Morheit) [reply src/DBConnection2.cpp]: @malirod , oops, seems I forgot to delete it. Unused `<iostream>` deleted

**Pull request closed without merging**